flash/data/process.py

```diff
@@ -286,25 +304,92 @@ def __init__(
         predict_transform: Optional[Dict[str, Callable]] = None,
+        save_meta: bool = True,
     ):
         super().__init__()
 
+        self.save_meta = save_meta
+
         # used to keep track of provided transforms
         self._train_collate_in_worker_from_transform: Optional[bool] = None
         self._val_collate_in_worker_from_transform: Optional[bool] = None
         self._predict_collate_in_worker_from_transform: Optional[bool] = None
         self._test_collate_in_worker_from_transform: Optional[bool] = None
 
-        self.train_transform = convert_to_modules(self._check_transforms(train_transform, RunningStage.TRAINING))
-        self.val_transform = convert_to_modules(self._check_transforms(val_transform, RunningStage.VALIDATING))
-        self.test_transform = convert_to_modules(self._check_transforms(test_transform, RunningStage.TESTING))
-        self.predict_transform = convert_to_modules(self._check_transforms(predict_transform, RunningStage.PREDICTING))
+        self._train_transform = self._check_transforms(train_transform, RunningStage.TRAINING)
+        self._val_transform = self._check_transforms(val_transform, RunningStage.VALIDATING)
+        self._test_transform = self._check_transforms(test_transform, RunningStage.TESTING)
+        self._predict_transform = self._check_transforms(predict_transform, RunningStage.PREDICTING)
 
-        if not hasattr(self, "_skip_mutual_check"):
-            self._skip_mutual_check = False
+        self.train_transform = convert_to_modules(self._train_transform)
+        self.val_transform = convert_to_modules(self._val_transform)
+        self.test_transform = convert_to_modules(self._test_transform)
+        self.predict_transform = convert_to_modules(self._predict_transform)
 
         self._callbacks: List[FlashCallback] = []
 
+        if self.save_meta:
+            self.state_dict = self._wrap_state_dict(self.state_dict)
+
+    def _wrap_state_dict(self, state_dict_fn: Callable) -> Callable:
+
+        @functools.wraps(state_dict_fn)
+        def wrapped_func(*args, **kwargs):
+            try:
+                from pip import get_installed_distributions
+            except:
+                from pip._internal.utils.misc import get_installed_distributions
+            try:
+                commit_sha = subprocess.check_output(["git", "rev-parse", "HEAD"]).decode("ascii").strip()
+            except:
+                commit_sha = "n/a"
+
+            try:
+                gitdiff = subprocess.check_output(["git", "diff"]).decode()
+            except:
+                gitdiff = ""
+            state_dict = state_dict_fn(*args, **kwargs)
+            state_dict["_meta"] = {}
+            state_dict["_meta"]["module"] = self.__module__
+            state_dict["_meta"]["class_name"] = self.__class__.__name__
+            state_dict["_meta"]["args"] = self._args
+            state_dict["_meta"]["kwargs"] = self._kwargs
```

Comment:
Should also add a `state_dict["_meta"]["_state"] = self._state` in order to preserve the labels (see here: https://github.com/PyTorchLightning/lightning-flash/blob/master/flash/data/process.py#L52)